Message forwarding (#6833)
hammerandchisel/discord-api-docs: docs/resources/Channel.md

| application_id?           | snowflake                                                                                                                                 | if the message is an [Interaction](#DOCS_INTERACTIONS_RECEIVING_AND_RESPONDING/) or application-owned webhook, this is the id of the application                                                                                                                        |
| flags?                    | integer                                                                                                                                   | [message flags](#DOCS_RESOURCES_CHANNEL/message-object-message-flags) combined as a [bitfield](https://en.wikipedia.org/wiki/Bit_field)                                                                                                                                 |

> | message_reference?        | [message reference](#DOCS_RESOURCES_CHANNEL/message-reference-structure) object                                                           | data showing the source of a crosspost, channel follow add, pin, or reply message                                                                                                                                                                                       |

Reviewer: ~~shouldn't this note forward too?~~ wait it was just for creation wasn't it

> | message_snapshots? \[5\]  | array of [message snapshot](#DOCS_RESOURCES_CHANNEL/message-snapshot-object) objects                                                      | the message associated with the message_reference. This is a minimal subset of fields in a message (e.g. `author` is excluded.)                                                                                                                                         |

Reviewer 2: ```suggestion
| message_snapshots? \[5\]  | array of [message snapshot](#DOCS_RESOURCES_CHANNEL/message-snapshot-object) objects                                                      | the message associated with the `message_reference`. This is a minimal subset of fields in a message (e.g. `author` is excluded.)                                                                                                                                         |
```

| type? \*            | integer   | [type of reference](#DOCS_RESOURCES_CHANNEL/message-reference-types).                                                                   |
| message_id?         | snowflake | id of the originating message                                                                                                           |
| channel_id? \*\*    | snowflake | id of the originating message's channel                                                                                                 |
| guild_id?           | snowflake | id of the originating message's guild                                                                                                   |
| fail_if_not_exists? | boolean   | when sending, whether to error if the referenced message doesn't exist instead of sending as a normal (non-reply) message, default true |

\* If `type` is unset, `DEFAULT` can be assumed in order to match the behaviour before message reference had types.
In future API versions this will become a required field.

> \*\* `channel_id` is optional when creating a reply, but will always be present when receiving an event/response that includes this data model.

Reviewer 2: If this is required (as per line 583) for forwards now, I think it would be good to put this info somewhere around here.

> \*\* `channel_id` is optional when creating a reply, but will always be present when receiving an event/response that includes this data model. Reqruired for forwards.

Reviewer 2: ```suggestion
\*\* `channel_id` is optional when creating a reply, but will always be present when receiving an event/response that includes this data model. **Required for forwards.**
```

- These are automatic messages sent when a message is pinned (type 6).
- These messages have `message_id` and `channel_id`, and `guild_id` if it is in a guild, which point to the message that was pinned.

###### Forwards

- These are messages which capture a snapshot of a message.

> - These messages have an array of [`message_snapshots`](#DOCS_RESOURCES_CHANNEL/message-snapshot-object) field containing a copy of the original message. This copy follows the same structure as a message, but has only the minimal set of fields returned required for context/rendering.

Reviewer 2: ```suggestion
- These messages have an array of [`message_snapshot`](#DOCS_RESOURCES_CHANNEL/message-snapshot-object) objects containing a copy of the original message. This copy follows the same structure as a message, but has only the minimal set of fields returned required for context/rendering.
```

  - of note: `author` will be excluded

> - A forwarded message can be identified by looking at it's `message_reference.type` field

Reviewer 2: ```suggestion
- A forwarded message can be identified by looking at its `message_reference.type` field
```

  - `message_snapshots` will be the message data associated with the forward. Currently we support only 1 snapshot.
  - prevents spoofing forwarded data

>   - `message_snapshots` are taken at moment the forward message is created, and are **immutable**; any mutations to the orignal message will not be propagated.

Reviewer 2: ```suggestion
  - `message_snapshots` are taken the moment a forward message is created, and are **immutable**; any mutations to the original message will not be propagated.
```

As of 2023-04-14, clients upload a 1 channel, 48000 Hz, 32kbps Opus stream in an OGG container.
The encoding, and the waveform details, are an implementation detail and may change without warning or documentation.

### Message Snapshot Object

#### Message Snapshot Structure

| Field      | Type      | Description                                                                      |
|------------|-----------|----------------------------------------------------------------------------------|
| message \* | [message] | subset of fields in the [message object](#DOCS_RESOURCES_CHANNEL/message-object) |

> | guild_id?  | snowflake | ID of the origin message's guild                                                 |

Reviewer 3: TODO: this has changed to 
```json
"guild": { "id": 1234 }
```

Reviewer 3: we've decided to actually not return this and instead direct people to use the `message.message_reference` of the parent

As of 2023-04-14, clients upload a 1 channel, 48000 Hz, 32kbps Opus stream in an OGG container.
The encoding, and the waveform details, are an implementation detail and may change without warning or documentation.

### Message Snapshot Object

#### Message Snapshot Structure

> | Field     | Type      | Description                                                                      |
> |-----------|-----------|----------------------------------------------------------------------------------|
> | message\* | [message] | subset of fields in the [message object](#DOCS_RESOURCES_CHANNEL/message-object) |

Reviewer 2: ```suggestion
| Field     | Type                                                             | Description                                       |
|-----------|------------------------------------------------------------------|---------------------------------------------------|
| message\* | partial [message](#DOCS_RESOURCES_CHANNEL/message-object) object | minimal subset of fields in the forwarded message |
```

> \* The current list of message fields subset consists of:

Reviewer 2: ```suggestion
\* The current subset of message fields consists of:
```

As of 2023-04-14, clients upload a 1 channel, 48000 Hz, 32kbps Opus stream in an OGG container.
The encoding, and the waveform details, are an implementation detail and may change without warning or documentation.

### Message Snapshot Object

#### Message Snapshot Structure

| Field     | Type      | Description                                                                      |
|-----------|-----------|----------------------------------------------------------------------------------|
| message\* | [message] | subset of fields in the [message object](#DOCS_RESOURCES_CHANNEL/message-object) |

\* The current list of message fields subset consists of:

> 'type`, `content`, `embeds`, `attachments`, `timestamp`, `edited_timestamp`, `flags`, `mentions`, `mention_roles`.

Reviewer 4: ```suggestion
`type`, `content`, `embeds`, `attachments`, `timestamp`, `edited_timestamp`, `flags`, `mentions`, `mention_roles`.
```

Reviewer 5: Is it intended that `mention_roles` is only returned on the Get Message(s) endpoints?

> \*\* This contains only the `id` for the guild. May hold additional fields in future apis.

Reviewer 2: ```suggestion
```

Post a message to a guild text or DM channel. Returns a [message](#DOCS_RESOURCES_CHANNEL/message-object) object. Fires a [Message Create](#DOCS_TOPICS_GATEWAY_EVENTS/message-create) Gateway event. See [message formatting](#DOCS_REFERENCE/message-formatting) for more information on how to properly format messages.

To create a message as a reply or forward of another message, apps can include a [`message_reference`](#DOCS_RESOURCES_CHANNEL/message-reference-structure).

> Refer to the documentation of for required fields.

Reviewer 2: ```suggestion
Refer to the documentation for required fields.
```
This sentence feels somewhat unnecessary...
